path/path.go

```diff
@@ -3,188 +3,301 @@ package path
+	Remainder() string
 }
 
-// Segments returns the different elements of a path
-// (elements are delimited by a /).
-func (p Path) Segments() []string {
-	cleaned := path.Clean(string(p))
-	segments := strings.Split(cleaned, "/")
+// ImmutablePath is a [Path] which is guaranteed to return "false" to [Mutable].
+type ImmutablePath struct {
+	Path
+}
 
-	// Ignore leading slash
-	if len(segments[0]) == 0 {
-		segments = segments[1:]
+func NewImmutablePath(p Path) (ImmutablePath, error) {
+	if p.Mutable() {
+		return ImmutablePath{}, fmt.Errorf("path was expected to be immutable: %s", p.String())
 	}
 
-	return segments
+	return ImmutablePath{p}, nil
+}
+
+type path struct {
+	str       string
+	root      cid.Cid
+	namespace string
+}
+
+func (p *path) String() string {
+	return p.str
+}
+
+func (p *path) Namespace() string {
+	return p.namespace
 }
 
-// String converts a path to string.
-func (p Path) String() string {
-	return string(p)
+func (p *path) Mutable() bool {
+	return p.namespace == IPNSNamespace
 }
 
-// IsJustAKey returns true if the path is of the form <key> or /ipfs/<key>, or
-// /ipld/<key>
-func (p Path) IsJustAKey() bool {
-	parts := p.Segments()
-	return len(parts) == 2 && (parts[0] == "ipfs" || parts[0] == "ipld")
+func (p *path) Root() cid.Cid {
+	return p.root
 }
 
-// PopLastSegment returns a new Path without its final segment, and the final
-// segment, separately. If there is no more to pop (the path is just a key),
-// the original path is returned.
-func (p Path) PopLastSegment() (Path, string, error) {
+func (p *path) Segments() []string {
+	// Trim slashes from beginning and end, such that we do not return empty segments.
+	str := strings.TrimSuffix(p.str, "/")
+	str = strings.TrimPrefix(str, "/")
 
-	if p.IsJustAKey() {
-		return p, "", nil
+	return strings.Split(str, "/")
+}
+
+type resolvedPath struct {
+	path
+	cid       cid.Cid
+	remainder string
+}
+
+func (p *resolvedPath) Cid() cid.Cid {
+	return p.cid
+}
+
+func (p *resolvedPath) Remainder() string {
+	return p.remainder
+}
+
+// NewIPFSPath returns a new "/ipfs" path with the provided CID.
+func NewIPFSPath(cid cid.Cid) ResolvedPath {
+	return &resolvedPath{
+		path: path{
+			str:       fmt.Sprintf("/%s/%s", IPFSNamespace, cid.String()),
+			root:      cid,
+			namespace: IPFSNamespace,
+		},
+		cid:       cid,
+		remainder: "",
 	}
+}
 
-	segs := p.Segments()
-	newPath, err := ParsePath("/" + strings.Join(segs[:len(segs)-1], "/"))
-	if err != nil {
-		return "", "", err
+// NewIPLDPath returns a new "/ipld" path with the provided CID.
+func NewIPLDPath(cid cid.Cid) ResolvedPath {
+	return &resolvedPath{
+		path: path{
+			str:       fmt.Sprintf("/%s/%s", IPLDNamespace, cid.String()),
+			root:      cid,
+			namespace: IPLDNamespace,
+		},
+		cid:       cid,
+		remainder: "",
 	}
+}
 
-	return newPath, segs[len(segs)-1], nil
+// NewIPNSPath returns a new "/ipns" path with the provided CID.
+func NewIPNSPath(cid cid.Cid) Path {
```

Comment:
@aschmahmann oh, yeah. I'm gonna change this to `ipns.Name`. I just had started this with cid and rebased it but the IPNS PR has landed 😄

Reply:
I removed IPNS and DNSLink related things for now. They can still be created manually, but we slightly limit the scope here.

Reply:
@lidel I added that. I think it can still be useful, as you can just `p := name.AsPath()`, instead of doing `p, err := path.NewPath(ipns.NamespacePrefix + name.String())` and then checking for the error.